types/react/index.d.ts

```diff
@@ -812,18 +818,62 @@ declare namespace React {
      * multiple sub-values. It also lets you optimize performance for components that trigger deep
      * updates because you can pass `dispatch` down instead of callbacks.
      *
-     * @version experimental
+     * @version 16.8.0
      * @see https://reactjs.org/docs/hooks-reference.html#usereducer
      */
-    function useReducer<S, A>(reducer: Reducer<S, A>, initialState: S, initialAction?: A | null): [S, Dispatch<A>];
+    // overload where "I" may be a subset of ReducerState<R>; used to provide autocompletion.
+    // If "I" matches ReducerState<R> exactly then the last overload will allow initializer to be ommitted.
+    // the last overload effectively behaves as if the identity function (x => x) is the initializer.
+    function useReducer<R extends Reducer<any, any>, I>(
+        reducer: R,
+        initializerArg: I & ReducerState<R>,
+        initializer: (arg: I & ReducerState<R>) => ReducerState<R>
+    ): [ReducerState<R>, Dispatch<ReducerAction<R>>];
+    /**
+     * An alternative to `useState`.
+     *
+     * `useReducer` is usually preferable to `useState` when you have complex state logic that involves
+     * multiple sub-values. It also lets you optimize performance for components that trigger deep
+     * updates because you can pass `dispatch` down instead of callbacks.
+     *
+     * @version 16.8.0
+     * @see https://reactjs.org/docs/hooks-reference.html#usereducer
+     */
+    // overload for free "I"; all goes as long as initializer converts it into "ReducerState<R>".
+    function useReducer<R extends Reducer<any, any>, I>(
+        reducer: R,
+        initializerArg: I,
+        initializer: (arg: I) => ReducerState<R>
+    ): [ReducerState<R>, Dispatch<ReducerAction<R>>];
+    /**
+     * An alternative to `useState`.
+     *
+     * `useReducer` is usually preferable to `useState` when you have complex state logic that involves
+     * multiple sub-values. It also lets you optimize performance for components that trigger deep
+     * updates because you can pass `dispatch` down instead of callbacks.
+     *
+     * @version 16.8.0
+     * @see https://reactjs.org/docs/hooks-reference.html#usereducer
+     */
+    // I'm not sure if I keep this 2-ary or if I make it (2,3)-ary; it's currently (2,3)-ary.
+    // The Flow types do have an overload for 3-ary invocation with undefined initializer.
+    // NOTE: the documentation or any alphas aren't updated, this is current for master.
+    // NOTE 2: without the ReducerState indirection, TypeScript would reduce S to be the most common
+    // supertype between the reducer's return type and the initialState (or the initializer's return type),
+    // which would prevent autocompletion from ever working.
+    function useReducer<R extends Reducer<any, any>>(
```

Comment:
Can this overload be 2-ary and be moved up to the top? Wouldn't that allow the compiler to catch the identity initializer?

If we do that, would there be a need for the other two overloads where `I` is a subset/supertype of the reducer state?, could we just only use the "free `I`" overload?

Reply:
Hmm.. testing it out on my local machine, I'm not seeing the issue with a setup like this:
```ts
type Dispatch<T> = (value: T) => void
type Reducer<S, A> = (state: S, action: A) => A
type State = {
  foo: string
  bar: boolean
}

declare const initialState: State
declare const initialArg: Pick<State, 'foo'>
declare function reducer(state: State, action: any): State

declare function useReducer<S, A>(reducer: Reducer<S, A>, initialState: S): [S, Dispatch<A>]
declare function useReducer<S, A, I>(
  reducer: Reducer<S, A>,
  initializerArg: I,
  initializer: (initializerArg: I) => S
): [ S, Dispatch<A> ]

// works fine, bar autocompleted, arg is correctly inferred
// S is inferred to still be the reducer return type (State), and not widened to the type of `initialArg`
useReducer(reducer, initialArg, arg => ({
  ...arg,
  bar: false
}))

// also works fine, arg  is correctly inferred
// S is inferred to still be the reducer return type
useReducer(reducer, initialState, arg => ({
  ...arg,
  bar: false
}))
```

Was the issue you were talking about that on typing out the second argument to set up the 3-ary overload, the compiler already starts complaining about the inassignability of the `I` generic to `S`? I thought that was just standard overloading behavior?